OpenRA.Mods.Common/Traits/Armament.cs

```diff
@@ -264,18 +262,45 @@ public virtual Barrel CheckFire(Actor self, IFacing facing, Target target)
 			var barrel = Weapon.Burst == 1 ? Barrels[currentBarrel] : Barrels[Burst % Barrels.Length];
 			currentBarrel++;
 
-			FireBarrel(self, facing, target, barrel);
-
+			FireBarrel(self, target, barrel);
 			UpdateBurst(self, target);
 
-			return barrel;
+			return true;
 		}
 
-		protected virtual void FireBarrel(Actor self, IFacing facing, Target target, Barrel barrel)
+		protected virtual void FireBarrel(Actor self, Target target, Barrel barrel)
 		{
 			foreach (var na in notifyAttacks)
 				na.PreparingAttack(self, target, this, barrel);
 
+			ScheduleDelayedAction(Info.FireDelay, () =>
+			{
+				var args = CreateProjectileArgs(self, target, barrel);
+				if (args == null)
+					return;
+
+				var projectile = args.Weapon.Projectile != null ? args.Weapon.Projectile.Create(args) : null;
+				if (projectile != null)
+					self.World.Add(projectile);
+
+				if (args.Weapon.Report != null && args.Weapon.Report.Any())
+					Game.Sound.Play(SoundType.World, args.Weapon.Report, self.World, self.CenterPosition);
+
+				if (Burst == args.Weapon.Burst && args.Weapon.StartBurstReport != null && args.Weapon.StartBurstReport.Any())
+					Game.Sound.Play(SoundType.World, args.Weapon.StartBurstReport, self.World, self.CenterPosition);
+
+				foreach (var na in notifyAttacks)
+					na.Attacking(self, target, this, barrel);
+
+				Recoil = Info.Recoil;
+			});
+		}
+
+		protected virtual ProjectileArgs CreateProjectileArgs(Actor self, Target target, Barrel barrel)
+		{
+			if (!CanFire(self, target, true))
```

Comment:
Consider a grenadier ordered to fire on a target that moves out of range during the FireDelay countdown. Does this mean that we will play the throwing sound and notify the attack, but not actually throw the projectile?

Reply:
Unless I'm misreading my own code, it shouldn't.

`var args` in the delayed action will be null if this check failed, so the delayed action should abort before the sounds and attack notification are triggered.